Add ANP Documentation

tssurya opened the pull request:
Add proper ANP docs for both end-users and developers

tssurya commented:
@astoycos and @danwinship @npinaeva PTAL

tssurya commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+  that as your infrastructure requirements evolve over time, you are able to insert
+  new ANPs when needed at the right priority level. If you pack your ANPs, then you
+  might need to recreate all of them to accommodate any changes in the future.
+* Each rule can have upto 100 peers. All those peers are added to the same address-set.
```
@igsilya : PTAL this point. Do you have other scale wise recommendations or bugs for ACLs/Address_Sets/PortGroups that we need to translate into OVNK world?

igsilya replied:
Are 'except' blocks allowed in ANP?  If so, I'd suggest discouraging from using these and suggest implementing the same rules with priorities, i.e. high priority rule to allow the IP block and a low priority rule to deny everything, instead of a single rule with an 'except' block.

tssurya replied:
no except blocks are not allowed in ANP, we can either do:

drop if cidr == A,B,C OR
allow if cidr == A,B,C

we cannot do "!="

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
-**OVN-Implementation:**
+* `subject` is the set of pods selected by `cluster-control` ANP, on
+  which the rules in the policy are applied on. Here its all the pods
+  in the cluster selected by the empty namespace selector.
```
```suggestion
  in the cluster.
```

(The way you wrote it can be read as "all of (the pods selected by the empty namespace selector)", suggesting that maybe some pods are not selected by that selector.)

tssurya replied:
yeah I see.. removed the redundant bit that causes the confusion.

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+* `subject` is the set of pods selected by `cluster-control` ANP, on
+  which the rules in the policy are applied on. Here its all the pods
+  in the cluster selected by the empty namespace selector.
+* Each AdminNetworkPolicy CRD will have a`.spec.priority` field. `priority`
```
```suggestion
* Each AdminNetworkPolicy resource will have a`.spec.priority` field. `priority`
```

A CRD is a Custom Resource _Definition_. The ANPs themselves are "resources" (or "custom resources" but there's usually no reason to clarify that they're "custom").

tssurya replied:
yeah thanks dan fixed it.

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
-The PortGroup for the above AdminNetworkPolicy is:
+Unlike ANP's BANP is a singleton object in your
+cluster meant to create a guardrail default security
+posture.
```
You can define terms like "guardrail" or "default security posture" such that they apply to BANP but not ANP, but unless you actually defined such a term, I don't think the ordinary meanings of those words clearly apply to BANP but not ANP.

I would say that BANP specifies baseline policies that apply unless they are overridden by an NP or ANP.

tssurya replied:
yeah agreed, fixed this and removed "guardrail" usage totally. what's in my head was an abstract non-defined concept I see.

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
````diff
@@ -1,303 +1,1118 @@
+    to:
+    - namespaces:
+        namespaceSelector: {}
 ```
````
You should probably explain this: that it blocks all traffic between all pods in the entire cluster by default, and requires everybody to create _both_ ingress and egress policies for all of the traffic that they want to allow.

Also, it applies even in private openshift namespaces that the cluster-admins don't know anything about, so probably you don't really want this policy...

tssurya replied:
yeah I have added a point in best practices bit to never use `{}` selectors but will clarify it here as well.

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+#### Pass Action: Delegate decision to NetworkPolicies
+
+In addition to setting `Deny` and `Allow` actions on ANP API rules,
+one can also use the `Pass` action for a rule. What this means is
```
(Note that you already used `Pass` in the examples before, without really explaining it.)

tssurya replied:
good point, I have left a comment on the above pass rules bit earlier on the pass rules (Please check the following sections to learn more about what `pass` is.)

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+network policies you don't need to create 1 policy per namespace.
+You need to only create 1 policy per tenant. Sample ANPs for
+multi-tenant isolation where we have 4 tenants `blue`, `green`,
+`yellow` and `red` are:
```
With these policies, all traffic within a tenant is forcibly Allowed; namespace admins are not permitted to have any within-tenant security. IMO, the more expected behavior for "multi-tenant isolation" would be a policy for `red` to `Deny` traffic not from `red`, a policy for `blue` to `Deny` traffic not from `blue`, etc, with no `Allow` rules.

(Admins almost never want to `Allow` arbitrary traffic. `Allow` in ANP does not mean "I want this allowed by default", it means "I want this to be allowed even if the user wants to block it". If you aren't explicitly trying to prevent users from blocking the traffic, then you mean `Pass`, not `Allow`.)

tssurya replied:
oh that's a great point!
so the expression of deny from everything else at prio 43 was meant to act like that deny rest of the traffic. So I have 5 ANPs at prio 40 that allow and then the ANP at prio 43 denies everything else.
So essentially its the same as "deny not from yourself" -> But I am also going to include your way of doing things.. that sounds like another second way to express it. So the way I had thought of this was allow whatever you want to allow at a higher prio and then deny everything else.

danwinship replied:
> So essentially its the same as "deny not from yourself"

No, it's not. It's "deny not from yourself _and you aren't allowed to deny anything else_".

The entire point of NP/ANP is that admins like to deny traffic. If pod A is not supposed to talk to pod B, then many admins want to be able to set policies so that pod A _can't_ talk to pod B.

But if you create an ANP that says "`Allow` from blue to blue", then Game Over. It is no longer possible to create an NP that says that pod A and pod B can't talk to each other, because the ANP would override it. So if a hacker can break into the most insecure pod in the tenant, then they can use that to reach every other pod in the tenant.

No cluster admin wants to create a policy that prevents project admins from being able to follow good security practices within their own namespaces.

tssurya replied:
sure if they want finer grained controls within the tenant they can, its all about the design ultimately, the requirements I portrayed here is "allow all pods in blue to be able to talk to all pods in blue" within them I don't want to restrict anything, so I can use allow.

If users want to put "within-tenant-further-control" then your scenario works and mine won't. I will document that.

tssurya replied:
I have clarified this point too, thanks Dan for raising this!

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+    Message:               Setting up OVN DB plumbing was successful
+    Reason:                SetupSucceeded
+    Status:                True
+    Type:                  Ready-In-Zone-ovn-control-plane
```
Oof... Having programmatically-generated Type names is not idiomatic kubernetes.

Note that `NetworkPolicy.Status` was abandoned because there didn't seem to be anything useful you could do with it

tssurya replied:
yeah this one of those IC zone things ultimately; we felt there was value in knowing which zone went wrong if any in applying the policy

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+  cluster including the infrastructure namespaces like kube-system and kapi-server
+  or openshift-*. Best practice to secure your workloads is to use workload
+  specific labels and then use that as the namespace selector instead.
+* `Pass` is a special case action used to delegate policy decision to the
+  network policy layer. If you don't care about interoperability with existing
+  network policies, then you probably don't need to worry about this feature.
```
disagree

tssurya replied:
have other suggestions on rewriting this? delegation is the only use case right?

npinaeva replied:
I think we can say that cluster admin **should** care about interoperability with existing network policies, if they have any. And clarify that if some ANP applies to the same connection as Network Policy, ANP's action will be enforced. If that is not desired, a higher-prio ANP with Skip action can be applied.

tssurya replied:
yeah I can rewrite it that way maybe it makes that more clear

danwinship replied:
As above, it's `Allow` that should be the rare special case, not `Pass`:

- "I want to block this traffic" → `Deny`
- "I don't want to block this traffic" → `Pass`
- "I want the pod to be forced to accept this traffic even if it doesn't want it" → `Allow`

tssurya replied:
but there is a section who don't want users to create network policies, they want to control it all with anps.. then they can just use allow or pass and given allow is allow in semantics its easier for them to do that.

tssurya replied:
I have tried to clarify this bit better in docs, let's see if it came out well.

danwinship replied:
> but there is a section who don't want users to create network policies

[citation required]

There are certainly lots of users who would explain what they want as "I want to deny x and y and z, and allow everything else", but I don't think any of them actually mean "`Deny` x and y and z, and `Allow` everything else".

danwinship replied:
anyway, whatever. We can wait and see what sort of questions/confusions we get from customers

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+## Known Limitations and Design Choices
+
+* Even if API supports upto 1000 as a value for `spec.priority`, **in OVN-Kubernetes
+  we only support upto 99 as a max value**. This is because of known limitation
```
Just to be clear, this is non-conformant, and if you don't already have a test for this in the upstream ANP tests, then you should add such a test so ovn-k can fail it.

If you feel that there is no way ovn-k could ever possibly implement this correctly then you should try to get the API changed upstream.

tssurya replied:
yeah I had raised this last year in u/s meeting and decision was to re-evaluate this at a later stage. The maximum OVN can do is upto 200 ANPs not more than that. I will bring this up in next u/s meeting.

tssurya replied:
also I have a test downstream where OVNK fails it. Probably should add one u/s as well -> just thought that given its a CNI problem maybe much better suited for d/s as a limitation

danwinship replied:
> The maximum OVN can do is up to 200 ANPs not more than that.

(The way _the current implementation_ works, it can't do more than 200 ANPs, but there are other possible implementations. AIUI, you currently reserve 10,000 OVN ACL priorities statically (e.g, `priority=23456` is the 56th rule of the ANP with priority 34 or whatever). This is efficient in terms of reprogramming OVN, but wasteful given that very very few ANPs are actually going to have 100 rules, so almost all of the priorities in the reserved range will go unused. A more dynamic mapping of rules to priorities would let you support more policies (in the normal case), though it would require occasionally renumbering existing ACLs to make more space...)

tssurya replied:
> > The maximum OVN can do is up to 200 ANPs not more than that.
> 
> (The way _the current implementation_ works, it can't do more than 200 ANPs, 

OVN can't do more than 200...

> but there are other possible implementations. AIUI, you currently reserve 10,000 OVN ACL priorities statically (e.g, `priority=23456` is the 56th rule of the ANP with priority 34 or whatever). This is efficient in terms of reprogramming OVN, but wasteful given that very very few ANPs are actually going to have 100 rules, so almost all of the priorities in the reserved range will go unused. A more dynamic mapping of rules to priorities would let you support more policies (in the normal case), though it would require occasionally renumbering existing ACLs to make more space...)

yes OVN-Kubernetes can change and tweak it dynamically on the fly, this was also discussed in past and deemed unclear on how useful this is reason being: we end up with another restriction like we support 1000 policies max with max 10 rules in each as worst case scenario? I guess we need data before doing that dynamic computing changes is how it ended IIRC. i.e do ppl prefer LOTS of policies with less rules in each OR do ppl prefer less policies with lots of rules in each; the problem is saying we support 400 or 500 or whatever that number is but we need to support it on original front with 100 rules on each policy even if users might or might not use it... 
I will open a tech debt item to come back to investigate this in the future based on user feedback in case ppl really need 600-700 policies or at least the ranges are needed..

tssurya replied:
I have clarified this point in the docs.

danwinship commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+  OVN-Kubernetes are from 0 to 99**. When user tries to create an ANP with priority
+  greater than 99 an event `ANPWithUnsupportedPriority` will be emitted. The user has
+  to manually update the priority field so that it get's created correctly.
+* The `subject` of an admin network policy does NOT include support for
+  `host-networked` pods.
+* The `namespaces` and `pods` peers in admin network policy does NOT include
+  support for `host-networked` pods.
+* `nodes` peer can be specified only from `egress` rule. There are no ingress use
+  cases yet which is why this is not supported from `ingress` rule.
+* `networks` peer can be specified only from `egress` rule. There are no ingress use
+  cases yet which is why this is not supported from `ingress` rule.
+* Specifying `namedPorts` with `networks` peer is not supported.
```
I think you should distinguish between "limitations/design choices of ovn-kubernetes's ANP implementation" and "limitations/design choices of the upstream API". (And maybe not even bother to list the latter.)

tssurya replied:
I separated it, its good to explicitly list some things out even for the API even if its already part of the CRD description fields..

danwinship reviewed (commented):
It's kind of a weird mix of end-user-focused and developer-focused...

tssurya commented:
> It's kind of a weird mix of end-user-focused and developer-focused...

yeah its meant to be both :) too much docs is ok

tssurya commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+administrators because:
+
+* They are not cluster-scoped and hence its hard to define a network of policies
+  that spawns across namespaces - you'd have to duplicate the same policy over and over
```
nice catch! :)

tssurya commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+* They cannot be created before the namespace is created (KAPI server expects the namespace
+  to be created first since it's namespace scoped); thus they cannot satisfy the requirements
+  where admins may want policies in the cluster to be in place before the workloads are
+  even created for security compliance and easier management.
```
true, gonna copy what you said :)

tssurya commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+  place when we create a policy and then we are expected to do an allowList of rules
+  in the policy. Then we make other exceptions like `except` blocks in `ipBlock` construct
+  which is an explicit deny within the allowList. Network administrators prefer to have
+  the power of defining what exactly to deny and allow instead of this implicit model.
```
I moved this point into a new paragraph and removed the deny bit and changed that to "what not to allow" bit

tssurya commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+There is a Feature Config option `enable-admin-network-policy` under
+`OVNKubernetesFeatureConfig` config that can be used to enable or disable this
+feature. When the feature is enabled, the CRDs are installed on the cluster
+and the `AdminNetworkPolicy` controller is invoked.
```
oh what a great call out. The feature is actually enabled by default :D, users can disable the feature using the flag. Let me ensure I write that well.

tssurya commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1118 @@
+        matchLabels:
+          tenant: restricted
+  - name: "default-deny" # rule4
     action: "Deny"
+    from:
+    - namespaces: {}
```
yeah this always trips me..

tssurya replied:
`    * rule4: deny all other ingress traffic` wrote it that way

and that is indeed a problem because we don't have a good way to say "allow from yourself"

tssurya commented:
I think I have fixed all Dan's comments except https://github.com/ovn-org/ovn-kubernetes/pull/4431/files#r1633267865
and I will bring https://github.com/ovn-org/ovn-kubernetes/pull/4431/files#r1633277714 up in u/s meeting

palonsoro commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1325 @@
+down to the specific policy in question you can grep for
+the name of the ANP you care about:
+
+* `ovn-nbctl list acl | grep cluster-control` (Do grep with -A, -B, -C to see the full ACL)
+* `ovn-nbctl list address-set | grep anp-name`
+* `ovn-nbctl list port-group | grep anp-name`
```
What about using `ovn-nbctl find` on the `external_ids` like this?
```
ovn-nbctl find ACL 'external_ids{>=}{"k8s.ovn.org/owner-type"=AdminNetworkPolicy,"k8s.ovn.org/name"=cluster-control}'

ovn-nbctl find Address_Set 'external_ids{>=}{"k8s.ovn.org/owner-type"=AdminNetworkPolicy,"k8s.ovn.org/name"=cluster-control}'

ovn-nbctl find Port_Group 'external_ids{>=}{"k8s.ovn.org/owner-type"=AdminNetworkPolicy,"k8s.ovn.org/name"=cluster-control}'
```

Even if the `ovn-nbctl list` is preferred, it would be good that if the ANP is called `cluster-control`, to use that name across the 3 commands.

tssurya replied:
yeah this works too, let me add this bit in.

tssurya replied:
done!

palonsoro commented on docs/features/network-security-controls/admin-network-policy.md:
````diff
@@ -1,303 +1,1325 @@
+translated into one or more ACLs like mentioned in the
+above section. Since there is no L4 match it should be
+translated to a single ACL. Knowing the rule number, we
+can grep for the correct ACL:
+
+`ovn-nbctl list acl | grep cluster-control:Egress:5`
+
+So `<anpName>:<direction>:<ruleNumber>` can be the grep
+pattern to get the right ACL. For this rule it looks like this:
+
+```yaml
+sh-5.2# ovn-nbctl list acl | grep cluster-control:Egress:5 -C 4
````
I posted this as a review by accident, but I have had problems in the past, so I deleted it and re-posted as a single comment:

Wouldn't it be cleaner to do this instead?
```
ovn-nbctl list ACL ANP:cluster-control:Egress:5
```

Part of the reason why I am suggesting to avoid `grep` in my recent suggestions is so that we can easily use the format-tweaking parameters of `ovn-nbctl` (e.g. `--format table`), which to me it is often helpful.

In addition, maybe we should have an example with the `<anpName>:<direction>:<ruleNumber>` in the command?

palonsoro replied:
Then, I guess it could be:
```
ovn-nbctl find ACL 'external_ids{>=}{"k8s.ovn.org/owner-type"=AdminNetworkPolicy,"k8s.ovn.org/name"=cluster-control,gress-index="5"}'
```

tssurya replied:
agreed that externalIDs is much better, will update it with that

tssurya replied:
done!

tssurya commented:
addressed @palonsoro 's comments and also addressed the two outstanding comments from Dan/Nadia around Pass and expression deny v/s allow in multi-tenancy

tssurya commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1351 @@
+  where admins may want the policies to be applicable to arbitrary future
+  namespaces that could get created in the cluster.
+
+Another motivating was to fix some of the usability aspects of the NetworkPolicy API.
```
```suggestion
Another motivation was to fix some of the usability aspects of the NetworkPolicy API.
```

tssurya commented:
@npinaeva PTAL I got some reviews from others on most of the things, but I still need a lgtm in the end :) So whenever you have some time would appreciate your reviews on this (last piece to finish that feature)

npinaeva commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1351 @@
 
+## Motivation
+
+NetworkPolicies in Kubernetes were created for developers to be able to secure their
```
should we add a link to the NetworkPolicies pointing to the API to clarify which object we mean here?

tssurya replied:
yep done.

npinaeva commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1351 @@
+  namespaces that could get created in the cluster.
+
+Another motivating was to fix some of the usability aspects of the NetworkPolicy API.
+The design of NetworkPolicy API is implicit which means the deny is already set in
+place when we create a policy and then we are expected to do an allowList of rules
```
This is a bit confusing to read, suggestion:
NetworkPolicy API is designed with an implicit deny-all policy, meaning that with the first created policy all traffic not allowed by it will be denied. NetworkPolicy only defines which connection will be allowed.

tssurya replied:
yeah good point, its amazing what things in my head become spaghetti when it comes out on paper. let me change this to your suggestion

danwinship replied:
> its amazing what things in my head become spaghetti when it comes out on paper.

nah, NetworkPolicy is nearly impossible to describe in a way that is both accurate and comprehensible. :grimacing:

tssurya replied:
fixed.

npinaeva commented on docs/features/network-security-controls/admin-network-policy.md:
```diff
@@ -1,303 +1,1351 @@
+Check out the original [User Stories](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2091-admin-network-policy#user-stories)
+for AdminNetworkPolicy. However note that that proposal is stale and it's better to
+follow the new developments via the [Network Policy Enhancement Proposals](https://network-policy-api.sigs.k8s.io/enhancements/).
+Let us look at some practical use cases of when to use admin network policies
```
Do you think we need the use cases here? Feels like it will become stale very soon, and we can just link to the upstream docs. (and do a better job upstream if the docs there are not good enough, but at least we will do that in one place)
Also, may be better to link user stories to https://network-policy-api.sigs.k8s.io/user-stories/ instead of the original KEP?

tssurya replied:
yeah I have changed the link to link to upstream website user stories.

npinaeva reviewed (commented):
great examples!

tssurya commented:
```
Summarizing 1 Failure:
  [FAIL] e2e egress firewall policy validation with external containers [It] Should validate the egress firewall policy functionality for allowed IP
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/egress_firewall.go:130

Ran 63 of 282 Specs in 2118.029 seconds
FAIL! -- 62 Passed | 1 Failed | 2 Pending | 217 Skipped
```
https://github.com/ovn-org/ovn-kubernetes/actions/runs/9846415399/job/27185200571?pr=4431
failed again

has the potential fix that went in:
```
dab68a2d6 (HEAD -> anp-docs, surya/anp-docs) Add ANP docs
34540a040 (origin/master, origin/HEAD, master) Merge pull request #4468 from flavio-fernandes/gh.workflow.release-1.0
66e08574d Merge pull request #4500 from npinaeva/lock-default-net
85c2db247 Merge pull request #4484 from npinaeva/e2e-ext-container
04fa7e437 Merge pull request #4498 from qinqon/ovnkube-trace-lane
85b08ee56 ovnkube-trace: Dummp empty tableid list on missing
e743946b5 udn, util: Add mgmt host iface name generator
aec6a9719 test, tools: Test dump-udn-vrf-table-ids
16169966f ci: Add lane to test ovnkube-trace
67735f1f2 Merge pull request #4458 from qinqon/net-segmentation-id-allocator
c8e2e2793 udn, allocator: Add masquerade IPs
2346f8eac Merge pull request #4503 from tssurya/ipgenerator-pkg
d3d5c9e98 (surya/ipgenerator-pkg, ipgenerator-pkg) Move ipgenerator to its own pkg
```
reopening flake

npinaeva reviewed (approved):
great work!

tssurya commented:
thanks Nadia, Dan, Pablo for reviews!
